File: app/src/main/java/me/liuyun/bjutlgn/db/DBHelper.java
package me.liuyun.bjutlgn.db;

import android.content.Context;
import android.database.sqlite.SQLiteDatabase;
import android.util.Log;

import com.j256.ormlite.android.apptools.OrmLiteSqliteOpenHelper;
import com.j256.ormlite.dao.Dao;
import com.j256.ormlite.support.ConnectionSource;
import com.j256.ormlite.table.TableUtils;

import java.sql.SQLException;

import me.liuyun.bjutlgn.entity.Flow;
import me.liuyun.bjutlgn.entity.User;

public class DBHelper extends OrmLiteSqliteOpenHelper {
    private static final String DATABASE_NAME = "bjutwifi.db";
    private static final int DATABASE_VERSION = 5;

    private Dao<Flow, Integer> flowDao = null;
    private Dao<User, Integer> userDao = null;

    public DBHelper(Context context) {
        super(context, DATABASE_NAME, null, DATABASE_VERSION);
    }

    @Override
    public void onCreate(SQLiteDatabase db, ConnectionSource connectionSource) {
        try {
            TableUtils.createTable(connectionSource, Flow.class);
            TableUtils.createTable(connectionSource, User.class);
        } catch (SQLException e) {
            Log.e(DBHelper.class.getName(), "onCreate", e);
            throw new RuntimeException(e);
        }
    }

    @Override
    public void onUpgrade(SQLiteDatabase db, ConnectionSource connectionSource, int oldVersion, int newVersion) {
        try {
            TableUtils.dropTable(connectionSource, Flow.class, true);
            TableUtils.dropTable(connectionSource, User.class, true);
            onCreate(db, connectionSource);
        } catch (SQLException e) {
            Log.e(DBHelper.class.getName(), "onUpgrade", e);
            throw new RuntimeException(e);
        }
    }

    public Dao<User, Integer> getUserDao() throws SQLException {
        if (userDao == null) {
            userDao = getDao(User.class);
        }
        return userDao;
    }

    public Dao<Flow, Integer> getFlowDao() throws SQLException {
        if (flowDao == null) {
            flowDao = getDao(Flow.class);
        }
        return flowDao;
    }

    @Override
    public void close() {
        super.close();
        flowDao = null;
        userDao = null;
    }
}

File: app/src/main/kotlin/me/liuyun/bjutlgn/db/DBHelper.kt
package me.liuyun.bjutlgn.db

import android.content.Context
import android.database.sqlite.SQLiteDatabase
import android.util.Log
import com.j256.ormlite.android.apptools.OrmLiteSqliteOpenHelper
import com.j256.ormlite.dao.Dao
import com.j256.ormlite.support.ConnectionSource
import com.j256.ormlite.table.TableUtils
import me.liuyun.bjutlgn.entity.Flow
import me.liuyun.bjutlgn.entity.User
import java.sql.SQLException

class DBHelper(context: Context) : OrmLiteSqliteOpenHelper(context, DBHelper.DATABASE_NAME, null, DBHelper.DATABASE_VERSION) {
    val flowDao: Dao<Flow, Int> by lazy { getDao<Dao<Flow, Int>, Flow>(Flow::class.java) }
    val userDao: Dao<User, Int> by lazy { getDao<Dao<User, Int>, User>(User::class.java) }

    override fun onCreate(db: SQLiteDatabase, connectionSource: ConnectionSource) {
        try {
            TableUtils.createTable(connectionSource, Flow::class.java)
            TableUtils.createTable(connectionSource, User::class.java)
        } catch (e: SQLException) {
            Log.e(DBHelper::class.java.name, "onCreate", e)
            throw RuntimeException(e)
        }

    }

    override fun onUpgrade(db: SQLiteDatabase, connectionSource: ConnectionSource, oldVersion: Int, newVersion: Int) {
        try {
            TableUtils.dropTable<Flow, Any>(connectionSource, Flow::class.java, true)
            TableUtils.dropTable<User, Any>(connectionSource, User::class.java, true)
            onCreate(db, connectionSource)
        } catch (e: SQLException) {
            Log.e(DBHelper::class.java.name, "onUpgrade", e)
            throw RuntimeException(e)
        }

    }

    companion object {
        private val DATABASE_NAME = "bjutwifi.db"
        private val DATABASE_VERSION = 5
    }
}
